src/lib/y2storage/proposal/space_maker.rb

```diff
@@ -173,23 +193,59 @@ def success?(planned_partitions)
       # @param lvm_helper [Proposal::LvmHelper] contains information about how
       #     to deal with the existing LVM volume groups
       # @param disk [String] optional disk name to restrict operations to
+      #
       def resize_and_delete!(planned_partitions, keep, lvm_helper, disk: nil)
-        log.info "Resize and delete. disk: #{disk}"
+        log.info "Resize and delete. disk: #{disk}, planned partitions:"
+        planned_partitions.each do |p|
+          log.info "  mount: #{p.mount_point}, disk: #{p.disk}, min: #{p.min}, max: #{p.max}"
+        end
 
+        # restart evaluation
         @distribution = nil
 
-        # Initially, resize only if there are no Linux partitions in the disk
-        resize_windows!(planned_partitions, disk, force: false)
-        delete_partitions!(planned_partitions, :linux, keep, disk)
-        delete_partitions!(planned_partitions, :other, keep, disk)
-        # If everything else failed, try resizing Windows before deleting it
-        resize_windows!(planned_partitions, disk, force: true)
-        delete_partitions!(planned_partitions, :windows, keep, disk)
-        # If deleting partitions was not enough, maybe there is no partition
-        # table and we have to wipe the disk
-        delete_disk_content(planned_partitions, lvm_helper, disk)
-
-        raise Error unless success?(planned_partitions)
+        # maybe it works already...
+        return if success?(planned_partitions)
+
+        # Try various methods to free space. Stop when a valid partition
+        # layout has been found.
+
+        methods_and_args = [
+          # step 1 - resize Windows partitions only if there are no Linux partitions on the disk
+          [:resize_windows!,     [{ force: false }]],
+          # step 2 - # delete Linux partitions
+          [:delete_partitions!,  [:linux, keep]],
+          # step 3 - delete other (non-Windows) partitions
+          [:delete_partitions!,  [:other, keep]],
+          # step 4 - resize Windows partitions
+          #
+          # ** Note **
+          #
+          # There are two steps where we try to resize a Windows partition.
+          # Both should be mutually exclusive. But the 'force' argument
```

Comment:
I don't get why this is a problem and why both must be mutually exclusive.

Reply:
There's no point in resizing the same partition twice. This led (in the broken code) to the strange situation where the second step tried to enlarge windows. In general, do the resize before or after deleting partitions, but not twice. I would have actually expected this force parameter to guarantee exactly this - but it doesn't.
I've added this comment to remind future developers what to expect here.

Reply:
But there is no resize operation happening twice in the disk or nothing like that. We are just changing the size of the windows region in the devicegraph in memory. As long as we don't call commit, I still don't see the problem with that.

Strictly speaking, it was not enlarging windows, it was just deciding that there was no need to reduce it as much as originally thought. That's not wrong. There is nothing like a real shrink and a real enlarge happening in the disk.